modulemd/v2/modulemd-module-stream.c

```diff
@@ -990,3 +990,15 @@ modulemd_module_stream_build_depends_on_stream (ModulemdModuleStream *self,
 
   return klass->build_depends_on_stream (self, module_name, stream_name);
 }
+
+    
+static gboolean
+modulemd_module_stream_default_equals (ModulemdModuleStream *self, ModulemdModuleStream *other)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_STREAM (self), FALSE);
+
+  guint64 mdversion_self = modulemd_module_stream_get_module_name (self);
+  guint64 mdversion_other = modulemd_module_stream_get_module_name (other);
+  g_return_val_if_fail ((mdversion_self == mdversion_other), FALSE);
```

Comment:
Don't use `g_return_val_if_fail()` for proper comparisons. This is actually an `assert()` under the hood. It's used in places like
```
g_return_val_if_fail (MODULEMD_IS_MODULE_STREAM (self), FALSE);
```
in order to catch programming mistakes. When compiled in debug mode, this will halt execution rather than returning FALSE. When compiled in release mode, it will log a loud error to the system log and return FALSE.

You want to do:
```
  if (mdversion_self != mdversion_other) return FALSE;
```

Reply:
Should I replace line 998 with an if statement that returns FALSE if self or other is NULL?

Reply:
No, that one is correct. That should be an assertion that a programmer did not accidentally pass in an incorrect variable type (like accidentally passing a `Modulemd.Module` instead of a `Modulemd.ModuleStream`).

The tests for equality are giving you the actual requested answer for this function: if they're not equal, the objects cannot be equal.

Oh, one other tip: include a test at the top of each equality function for whether `self == other`. If, for any reason, you've been passed the exact same pointer address, you can short-circuit all of the other comparisons and return TRUE, since it's not just equivalent: it's the same location in memory.

Reply:
I'm not sure if I was clear enough above, so let me try to explain. The purpose of `g_return_val_if_fail()` is that it is intended for non-standard reasons to exit. Essentially, it's a side-effect. We call this when we're making sure that the person who wrote the code did not make a mistake with what they passed in.

We want to use standard return values for any situation where the return value is the correct answer to whatever question the function is supposed to answer.

In this case, the function is supposed to answer: "Do these two `Modulemd.ModuleStream` instances have equivalent values for all of their attributes?". If we're comparing their mdversions or their stream name, that's answering that question. If, however, we're just making sure that we were passed real `Modulemd.ModuleStream` objects, that's not answering the same question. So we treat that as a non-standard error and use `g_return_val_if_fail()`